### PR TITLE
Schema addChildCheck/addAttributeCheck filtering

### LINKED ISSUE #15834: `Schema#addChildCheck` leads to performance problems
## 📝 Provide a description of the improvement

Currently, we can use `Schema#addChildCheck()` to add custom callback that can affect whether some node is allowed at given place in model structure. The callbacks were meant to be flexible, you can check anything and as you wish in it. Same for `Schema#addAttributeCheck()`.

Unfortunately, this can lead to performance problems if you start using many of them. Why?

Schema is checked very frequently by multiple features. Most commands check either whether it can insert a child at given place or whether it is possible to apply attribute for given selection. This is refreshed after every selection change and model change. With a rich set of plugins, you may have as many as 10-15 checks every keystroke. When multiple callbacks like this are added, this can become 50-100 callbacks each keystroke, and you may start to see downgrade in the editor responsiveness, especially on slower machines.

This is of course not necessary. There's no need for features A,B,C,D... to check whether feature X complies with schema. But since these are very simple callbacks, they are fired each time the schema checked by each feature.

However, we use declarative schema so much, how come it does not affect performance?

Well, declarative schema is checked only for a specified element/attribute. `Schema#checkChild` and `Schema#checkAttribute` gets a definition for specified element/attribute and only checks it. As a result feature X checks only feature X rules.

We should try to limit how many custom callbacks are called after each model update.

First, we should add additional parameter to `addChildCheck()` and `addAttributeCheck()` that will specify for which element/attribute this callback is declared. We can have this optional to still allow for very general checks but we would assume that these are used very rarely, for complex custom plugins, and maybe there's one or two such callbacks in the editor at most.

We should try to keep `checkChild` and `checkAttribute` events for backward compatibility but it might be that the arguments for these events will change, which will affect all `addChildCheck()` uses. But... as I look at it know, it shouldn't change, as we normalize the arguments in the highest priority listener:

```js
this.on( 'checkChild', ( evt, args ) => {
	args[ 0 ] = new SchemaContext( args[ 0 ] );
	args[ 1 ] = this.getDefinition( args[ 1 ] );
}, { priority: 'highest' } );
```

So, after adding the extra parameter, we have two ways of solving the performance issue.

First -- instead of simply adding the callback as an event listener, we would save it in an internal map in `Schema`. Then, when `checkChild()` is called, before doing default declarative check, we would go through all the registered callbacks for given definition.

Second -- we could use namespaced events. So, if I check schema for `paragraph`, then `checkChild:paragraph` is called. In this solution, we don't need to change much how `addChildCheck()` works. We will simply add callbacks to the namespaced events. However, `checkChild()` cannot be a decorated method any longer. Instead, it would have to be like this:

```js
public checkChild( context, def ) {
    normalize(); // What happens in highest priority listener now.

    if ( !def ) {
        return false; // As in `checkChild()` now.
    }

    // The default check in `_checkContextMatch()` must be defined as `low` priority callback.
    this.fire( 'checkChild:' + def.name );
}
```

I am fine with both solutions. One thing that I don't like about the second solution a bit is that we have a method (`childCheck()`) which behaves much like a decorated method but isn't really, which may be confusing? But this is a tiny nitpick.

### PR DESCRIPTION
**Pull request by map-r**
### Suggested merge commit message ([convention](https://ckeditor.com/docs/ckeditor5/latest/framework/contributing/git-commit-message-convention.html))

Feature (engine): `Schema#addChildCheck()` and `Schema#addAttributeCheck()` can now register a callback for specific item or attribute, which should improve performance when using custom callback checks. Callback checks should be added only for specific item or attribute if possible. See the API reference. Closes #15834.

Fix (engine): `Schema#checkChild()` will now correctly check custom callback checks for each item in the context.

Other (engine): Introduced `Schema#trimLast()`.

Internal: Used specific callback checks in some of the features.

Docs (engine): Improved schema deep dive guide and added missing section related to attributes custom callback checks.

MINOR BREAKING CHANGE: `Schema` callbacks added through `addChildCheck()` will no longer add event listeners with `high` priority and will no longer stop `checkChild` event. Instead, these callbacks are now handled on `normal` priority, as a part of the default `checkChild()` call. This also means that listeners added to `checkChild` event on `high` priority will fire before any callbacks added by `checkChild()`. Earlier they would fire in registration order. This may impact you if you implemented custom schema callback using both `addChildCheck()` and direct listener to `checkChild` event. All above is also true for `addAttributeCheck()` and `checkAttribute` event and callbacks.

---

### Additional information

### CODE REVIEW

File: packages/ckeditor5-engine/src/model/schema.ts
Comment (scofalik) [reply packages/ckeditor5-engine/src/model/schema.ts:634]: Although I think I understand why you went this way. We have two types of callbacks when it comes to current `addAttributeCheck`.

First -- we have custom callbacks for attribute X. They would be solved if you implemented it accordingly to my idea (set callback for attribute name, not element name).

But then, we have attribute-related callbacks for element Y. For example, disallow all attributes in `$text` inside `title-content`. But in this case, such callback should be actually called for all attributes, so it makes sense that it would stay as a generic callback. It's the same as adding one callback for all attributes.

OTOH, if a callback stays generic it will be also called for non-related items. Let's take example rule:

```javascript
schema.addAttributeCheck( ( context, attributeName ) => {
	if ( context.endsWith( 'codeBlock $text' ) ) {
		return false;
	}
} );
```

Now it would be fired for, e.g. `schema.checkAttribute( [ 'image' ], 'imageSrc' )`, which doesn't make much sense. If we specify `$text` here, and only get callbacks for `context.last.name` and `attributeName`, then we could limit the number of callbacks even more.

**Review by scofalik — CHANGES_REQUESTED** [review]
I added some comments.

One decision to make here is whether we want to stick to the original plan (my advice to keep things simple) or if we want to extend the original plan a bit, following what @map-r did in this PR.

In short (read my comments to understand more):

1. The original was to have `addChildCheck( callback, elementName )` and `addAttributeCheck( callback, attributeName )` so that these callbacks are fired only for given element name (`checkChild( context, elementName )`) or attribute (`checkAttribute( context, attributeName )`).
2. However they could be extended to cover more cases.
3. It would be something like `addChildCheck( callback, elementOrParentName )` and `addAttributeCheck( callback, attributeOrElementName )`.
4. With these extensions, it seems we would cover most (all?) of the cases we have right now. Callback added by both methods would be additionally fired if `context.last.name` matches the parameter.
5. I'd advise closing the other PR (introducing disallowing in schema) first, and then rewrite some of the callbacks to static declarations. And then see how many more callbacks we need to cover.
6. In general, I'd err on keeping this simple. If there's one or two global callbacks, then we are fine. We can always extend later. OTOH, note, that if you go with extensions how Adam proposed, you need to make extra checks in the callbacks (which I commented on in this review -- I thought they are not needed but they actually are). Extending later would be a breaking change

@DawidKossowski @map-r @niegowski I leave this to you.

File: packages/ckeditor5-engine/src/model/model.ts
Comment (map-r) [reply packages/ckeditor5-engine/src/model/model.ts:152]: we run it for both $marker as item in context and as a child element

File: packages/ckeditor5-engine/src/model/schema.ts
Comment (map-r) [reply packages/ckeditor5-engine/src/model/schema.ts:70]: I think not, just moved the checkChild listener to here and made it run through all former listeners (now functions stored in map)

File: packages/ckeditor5-engine/src/model/model.ts
Comment (map-r) [reply packages/ckeditor5-engine/src/model/model.ts:152]: thus, it might fire for wrong thing (when $marker is in context and something else is child)

File: packages/ckeditor5-engine/src/model/schema.ts
Comment (map-r) [reply packages/ckeditor5-engine/src/model/schema.ts:70]: And actually I understood it like so and it seemed safer not changing the priorities. I'll try another approach which you suggested

File: packages/ckeditor5-engine/src/model/schema.ts
Comment (map-r) [reply packages/ckeditor5-engine/src/model/schema.ts:80]: everything we would need at this point is to make a filter object, which either has `context` or a `child` as a property, with the name of element being the value. Then we will only check what we want only once, as calling this twice is what you say you have problem with.

File: packages/ckeditor5-engine/src/model/schema.ts
Comment (map-r) [reply packages/ckeditor5-engine/src/model/schema.ts:634]: I guess making two properties to filter in both childCheck and attributeCheck seems reasonable. I can leave it like this to be improved, or try to improve it by this two-type filtering, tbd.

File: packages/ckeditor5-engine/src/model/schema.ts
Comment (map-r) [reply packages/ckeditor5-engine/src/model/schema.ts]: The above comment as of today is still to be resolved, but probably in sync with that other PR Szymon mentioned

File: packages/ckeditor5-code-block/src/codeblockediting.ts
Comment (scofalik) [reply packages/ckeditor5-code-block/src/codeblockediting.ts:158]: Outdated after merging the other PR.

**Comment by scofalik** [comment]
Summary:

1.  `Schema#addChildCheck()` and `Schema#addAttributeCheck()` have a new optional parameter that allows to register the callback for a specific item or specific attribute.
2.  Earlier, when you registered a custom callback, it was fired for every `checkChild()` and `checkAttribute()` call. With multiple features and multiple custom callbacks, it could lead to performance problems (lagging when typing).
3.  The way how it works is very simple:
    1.  If you registered `addChildCheck( cb, 'foo' )` it will be called when `checkChild( ctx, 'foo' )` is called.
        1.  Mind, that `checkChild( ctx )` also checks the whole `ctx`. If `'foo'` is inside `ctx` the callback will be fired too.
    2.  If you registered `addAttributeCheck( cb, 'bold' )` it will be called when `checkAttribute( ctx, 'bold' )` is called.
        1.  Mind, that you register attribute check for a model attribute not a model item.
4.  Generic callbacks are still available.
    1.  Of course, generic callbacks are not recommended as they can lead to performance problems. However, sometimes, they can't be avoided.
5.  `checkChild()` and `checkAttribute()` are still decorated so you can add listeners to the events. This is not recommended as it can also lead to performance problems.
6.  The order of performing checks is as follows:
    1.  All callbacks added as events listeners with `highest` and `high` priority.
    2.  All generic callbacks. Among them, order as added in code.
    3.  All specific callbacks. Among them, order as added in code.
    4.  Declarative checks.

I also added this summary to the issue so it can be easy to found.

File: packages/ckeditor5-engine/tests/model/schema.js
Comment (niegowski) [inline packages/ckeditor5-engine/tests/model/schema.js:821]: Shouldn't this be marked as a breaking change? Same as for the attribute check?
What if someone provided a child check callback or a custom listener who relied on the event being stopped before?

**Review by niegowski — CHANGES_REQUESTED** [review]

File: packages/ckeditor5-engine/src/model/schema.ts
Comment (scofalik) [reply packages/ckeditor5-engine/src/model/schema.ts:460]: It gets 100% CC.

File: packages/ckeditor5-engine/src/model/schema.ts
Comment (scofalik) [reply packages/ckeditor5-engine/src/model/schema.ts:583]: It gets 100% CC.

File: packages/ckeditor5-engine/src/model/schema.ts
Comment (scofalik) [reply packages/ckeditor5-engine/src/model/schema.ts:655]: It gets 100% CC.

File: packages/ckeditor5-engine/src/model/schema.ts
Comment (scofalik) [reply packages/ckeditor5-engine/src/model/schema.ts:1069]: It gets 100% CC.

File: packages/ckeditor5-engine/src/model/schema.ts
Comment (scofalik) [reply packages/ckeditor5-engine/src/model/schema.ts:1869]: I used `pop()` to have analogy with `push()` method that is available on the `SchemaContext` already. I don't like it either. I don't like `trim()` and `trimEnd()` as it returns "empty" values and "end" is uncountable (how much end, how many items?). But I like `trimLast()` it specifically tells you that it will trim the last element.

File: packages/ckeditor5-engine/tests/model/schema.js
Comment (scofalik) [reply packages/ckeditor5-engine/tests/model/schema.js:821]: I tried to analyze this and you are right, and I added a minor breaking change entry.

However, if I get it correctly, the scope for possible problem is quite narrow here.

If somebody was using only direct event listeners (i.e. `schema.on( 'checkChild' )`) then it seems to me that nothing changed for them. They were probably adding these events on `high` priority and were stopping them (otherwise, these callback would not make much sense, I think).

If somebody was using only callbacks added through `addChildCheck()` or `addAttributeCheck()` then nothing changes here as well -- they are still working in "first wins" fashion, even though the event is not stopped.

So if I get it right, the only problem may happen if somebody was defining rules both using the event listener and using the methods, because they might have first define `addChildCheck( callback )` and then add listener with `high` priority, and in such case, if `callback` returned a value, the listener would not fire. But now, not only the listener will fire, it will fire before the callback.

Comment (scofalik) [reply packages/ckeditor5-engine/tests/model/schema.js:821]: Let me know what you think about the entry. I first described it for "child checks" and then at the end added that it is true also for "attribute checks". When I was describing both at once, it became almost unreadable.

**Review by scofalik — DISMISSED** [review]

**Review by niegowski — APPROVED** [review]

**Pull request merged**
